File: src/main/java/org/apache/spark/launcher/KyuubiMain.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.launcher;

import java.util.*;

import static org.apache.spark.launcher.CommandBuilderUtils.*;

public class KyuubiMain {
    /**
     * Usage: Main [class] [class args]
     */
    public static void main(String[] argsArray) throws Exception {
        checkArgument(argsArray.length > 0, "Not enough arguments: missing class name.");

        List<String> args = new ArrayList<>(Arrays.asList(argsArray));
        String className = args.remove(0);

        AbstractCommandBuilder builder;
        try {
            builder = new SparkSubmitCommandBuilder(args);
        } catch (IllegalArgumentException e) {
            System.err.println("Error: " + e.getMessage());
            System.err.println();

            MainClassOptionParser parser = new MainClassOptionParser();
            try {
                parser.parse(args);
            } catch (Exception ignored) {
                // Ignore parsing exceptions.
            }

            List<String> help = new ArrayList<>();
            if (parser.className != null) {
                help.add(parser.CLASS);
                help.add(parser.className);
            }
            help.add(parser.USAGE_ERROR);
            builder = new SparkSubmitCommandBuilder(help);
        }

        Map<String, String> env = new HashMap<>();
        List<String> cmd = builder.buildCommand(env);
        System.err.println("Spark Command: " + join(" ", cmd));
        System.err.println("========================================");

        // In bash, use NULL as the arg separator since it cannot be used in an argument.
        List<String> bashCmd = prepareBashCommand(cmd, env);
        for (String c : bashCmd) {
            System.out.print(c);
            System.out.print('\0');
        }
    }

    /**
     * Prepare the command for execution from a bash script. The final command will have commands to
     * set up any needed environment variables needed by the child process.
     */
    private static List<String> prepareBashCommand(List<String> cmd, Map<String, String> childEnv) {
        if (childEnv.isEmpty()) {
            return cmd;
        }

        List<String> newCmd = new ArrayList<>();
        newCmd.add("env");

        for (Map.Entry<String, String> e : childEnv.entrySet()) {
            newCmd.add(String.format("%s=%s", e.getKey(), e.getValue()));
        }
        newCmd.addAll(cmd);
        return newCmd;
    }

    /**
     * A parser used when command line parsing fails for spark-submit. It's used as a best-effort
     * at trying to identify the class the user wanted to invoke, since that may require special
     * usage strings (handled by SparkSubmitArguments).
     */
    private static class MainClassOptionParser extends SparkSubmitOptionParser {

        String className;

        @Override
        protected boolean handle(String opt, String value) {
            if (CLASS.equals(opt)) {
                className = value;
            }
            return false;
        }

        @Override
        protected boolean handleUnknown(String opt) {
            return false;
        }

        @Override
        protected void handleExtraArgs(List<String> extra) {

        }

    }
}

File: kyuubi-server/src/main/scala/org/apache/spark/launcher/KyuubiMain.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.spark.launcher

import java.util.{ArrayList => JAList, HashMap => HMap, List => JList, Map => JMap}

import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer
import scala.util.{Failure, Success, Try}

import org.apache.spark.launcher.CommandBuilderUtils._

object KyuubiMain {

  @throws[Exception]
  def main(args: Array[String]): Unit = {

    val argAsJava = args.toList.asJava
    val printStream = System.err
    val builder = Try { new KyuubiSubmitCommandBuilder(argAsJava) } match {
      case Success(b) => b
      case Failure(e) =>
        // scalastyle:off
        printStream.println("Error: " + e.getMessage)
        printStream.println()
        // scalastyle:on
        val parser = new MainClassOptionParser
        try {
          parser.parse(argAsJava)
        } catch {
          case _: Throwable => // ignored
        }

        val help = new ArrayBuffer[String]
        if (parser.className != null) {
          help += parser.CLASS
          help += parser.className
        }
        help += parser.USAGE_ERROR
        new SparkSubmitCommandBuilder(help.toList.asJava)
    }

    val env = new HMap[String, String]()
    val cmd = builder.buildCommand(env)
    // scalastyle:off
    printStream.println("Kyuubi Command: " + join(" ", cmd))
    printStream.println("========================================")
    // scalastyle:on
    prepareBashCommand(cmd, env).asScala.foreach { c =>
      print(c)
      print('\u0000')
    }
  }

  private[this] def prepareBashCommand(
      cmd: JList[String],
      childEnv: JMap[String, String]): JList[String] = {
    if (childEnv.isEmpty) return cmd
    val newCmd = new JAList[String]
    newCmd.add("env")
    childEnv.asScala.foreach(e => newCmd.add(String.format("%s=%s", e._1, e._2)))
    newCmd.addAll(cmd)
    newCmd
  }

  private class MainClassOptionParser extends SparkSubmitOptionParser {
    private[launcher] var className: String = _
    override protected def handle(opt: String, value: String): Boolean = {
      if (CLASS == opt) {
        className = value
      }
      false
    }
    override protected def handleUnknown(opt: String) = false
    override protected def handleExtraArgs(extra: JList[String]): Unit = {}
  }

}